File: src/lib/attestation.ts
import { X509Certificate, X509ChainBuilder } from "@peculiar/x509";
import { decode, encode } from "@stablelib/base64";
import * as cbor from "cbor2";
import { z } from "zod";
import { fetchAttestationDocument, getApiUrl } from "./api";
import awsRootCertDer from "../assets/aws_root.der";

// Assert that the root cert is not empty
if (!awsRootCertDer || awsRootCertDer.length === 0) {
  throw new Error("AWS root certificate is empty or not loaded correctly");
}

const AttestationDocumentSchema = z.object({
  module_id: z.string().min(1),
  digest: z.literal("SHA384"),
  timestamp: z.number().min(1677721600),
  pcrs: z.map(z.number(), z.instanceof(Uint8Array)),
  certificate: z.instanceof(Uint8Array),
  cabundle: z.array(z.instanceof(Uint8Array)),
  public_key: z.nullable(z.instanceof(Uint8Array)),
  user_data: z.nullable(z.instanceof(Uint8Array)),
  nonce: z.nullable(z.instanceof(Uint8Array))
});

export type AttestationDocument = z.infer<typeof AttestationDocumentSchema>;

const ParsedAttestationDocumentSchema = z.object({
  protected: z.instanceof(Uint8Array),
  // There's an "unprotected" header in the CBOR, but we never use it
  payload: z.instanceof(Uint8Array),
  signature: z.instanceof(Uint8Array)
});

type ParsedAttestationDocument = z.infer<typeof ParsedAttestationDocumentSchema>;

export async function parseDocumentData(
  attestationDocumentBase64: string
): Promise<ParsedAttestationDocument> {
  try {
    if (!attestationDocumentBase64) {
      throw new Error("Attestation document is empty.");
    }
    const attestationDocumentBuffer = decode(attestationDocumentBase64);
    const cborAttestationDocument: Uint8Array[] = cbor.decode(attestationDocumentBuffer);
    const protectedHeader = cborAttestationDocument[0];

    // There's an "unprotected" header in the CBOR, but we never use it
    // const unprotectedHeader = cborAttestationDocument[1];

    // The payload is another CBOR thing but we parse it later
    const payload = cborAttestationDocument[2];
    const signature = cborAttestationDocument[3];

    // Zod will make sure these are all Uint8Arrays
    // TODO: validate length?
    const zodParsed = ParsedAttestationDocumentSchema.parse({
      protected: protectedHeader,
      payload,
      signature
    });
    return zodParsed;
  } catch (error) {
    console.error("Error parsing document data:", error);
    throw new Error("Failed to parse document data.");
  }
}

// Parses the internal attestation document
export async function parseDocumentPayload(payload: Uint8Array): Promise<AttestationDocument> {
  try {
    const documentData = cbor.decode(payload);
    const zodParsed = AttestationDocumentSchema.parse(documentData);
    return zodParsed;
  } catch (error) {
    console.error("Error parsing document payload:", error);
    throw new Error("Failed to parse document payload.");
  }
}

// Creates and serializes a COSE Signature1 structure exactly matching the Rust implementation
export function createSigStructure(bodyProtected: Uint8Array, payload: Uint8Array): Uint8Array {
  // This exactly matches the Rust SigStructure::new_sign1 implementation
  const sig1 = [
    "Signature1", // Context string
    bodyProtected, // Protected headers
    new Uint8Array(0), // external_aad (empty ByteBuf in Rust)
    payload // payload
  ];

  // Encode to CBOR - equivalent to Rust's as_bytes()
  return cbor.encode(sig1);
}

// Creates a COSE_Sign1 thing and verifies its signature
async function verifySignature(
  documentData: ParsedAttestationDocument,
  publicKey: CryptoKey
): Promise<boolean> {
  try {
    console.log("SIGNATURE:");
    console.log(encode(documentData.signature));

    // Create signature bytes
    const signatureBytes = createSigStructure(documentData.protected, documentData.payload);

    // Hash the COSE_Sign1 with SHA-384 for debugging (crypto subtle will hash it as well);
    const hashBuffer = await crypto.subtle.digest("SHA-384", signatureBytes);

    console.log("SIGNATURE STRUCTURE DIGEST:");
    console.log(encode(new Uint8Array(hashBuffer)));

    // Verify the signature using Web Crypto API
    const verified = await crypto.subtle.verify(
      // TODO: these could be derived from the document, but we're hardcoding them for now
      {
        name: "ECDSA",
        hash: "SHA-384"
      },
      publicKey,
      documentData.signature,
      signatureBytes
    );

    return verified;
  } catch (error) {
    console.error("Error verifying signature:", error);
    if (error instanceof Error) {
      throw new Error(`Signature verification failed: ${error.message}`);
    } else {
      throw new Error(`Signature verification failed: ${error}`);
    }
  }
}

export async function authenticate(
  attestationDocumentBase64: string,
  trustedRootCert: Uint8Array,
  nonce: string
): Promise<AttestationDocument> {
  try {
    // Following the steps here: https://docs.aws.amazon.com/enclaves/latest/user/verify-root.html
    // Step 1. Decode the CBOR object and map it to a COSE_Sign1 structure
    const parsedDocument = await parseDocumentData(attestationDocumentBase64);

    // Step 2. Extract the attestation document from the COSE_Sign1 structure
    const document = await parseDocumentPayload(parsedDocument.payload);

    // Step 2.5 Make sure the nonce matches
    if (!document.nonce) {
      throw new Error("Attestation document does not have a nonce.");
    }

    const decoder = new TextDecoder("utf-8");
    const documentNonce = decoder.decode(document.nonce);

    if (nonce !== documentNonce) {
      console.log("Nonce mismatch");
      console.log("Provided nonce:", nonce);
      console.log("Attestation document nonce:", documentNonce);
      throw new Error("Attestation document's nonce does not match the provided nonce.");
    }

    // Step 3. Verify the certificate's chain
    const certs = [];

    // Verify that the trusted root cert is the same as the first cert in the cabundle
    const firstCertBase64 = encode(document.cabundle[0]);
    if (firstCertBase64 !== encode(trustedRootCert)) {
      console.error("Root cert doesn't match first cert");
      console.log("First cert base64:", firstCertBase64);
      console.log("Trusted root cert base64:", encode(trustedRootCert));
      throw new Error("Root cert does not match first cert in attestation document.");
    }

    // Add the cabundle to the certs array... rust does this in reverse but it doesn't seem to matter
    for (let i = 0; i < document.cabundle.length; i++) {
      const cert = new X509Certificate(document.cabundle[i]);
      certs.push(cert);
    }

    // Add the document's cert as the "leaf" cert (the one that we're verifying)
    const leafCert = new X509Certificate(document.certificate);

    const chain = new X509ChainBuilder({
      certificates: certs
    });

    // `.build` checks each cert in the chain that it verifies with a pubkey of another cert in the chain
    // and only adds the ones that pass. If there's a failure there will be a length mismatch.
    // https://github.com/PeculiarVentures/x509/blob/1158bcbe2c5393196d66b9096b7b5e9c35901bce/src/x509_chain_builder.ts#L61
    const certChainItems = await chain.build(leafCert);
    console.log("Chain items:", certChainItems);

    // The chain builder checks signatures but not expiration
    // So let's check for expiration ourselves
    const date = new Date();
    const time = date.getTime();
    for (let i = 0; i < certChainItems.length; i++) {
      const cert = certChainItems[i];
      console.log("CERT: ", i);
      console.log(cert.subject);
      console.log("Not before:", cert.notBefore);
      console.log("Not after:", cert.notAfter);
      console.log(cert.toString("pem"));

      if (cert.notBefore.getTime() > time || cert.notAfter.getTime() < time) {
        throw new Error("Certificate is expired.");
      } else {
        console.log(`Certificate ${i} is not expired.`);
      }
    }

    // The chain should have the whole cabundle plus the leaf cert
    if (certChainItems.length !== document.cabundle.length + 1) {
      throw new Error("Certificate chain length does not match length of cabundle.");
    }

    // Step 4. Ensure the attestation document is properly signed

    // Get the webcrypto version of the public key (the raw bytes are the same fwiw)
    const publicKey = leafCert.publicKey;
    console.log("PUBLIC KEY:");
    console.log(encode(new Uint8Array(publicKey.rawData)));
    const pubEcKey = await publicKey.export();

    // Verify the signature!
    const verified = await verifySignature(parsedDocument, pubEcKey);

    console.log("Signature verified:", verified);

    if (!verified) {
      throw new Error("Signature verification failed.");
    }

    return document;
  } catch (error) {
    console.error("Error verifying attestation document:", error);
    throw error;
  }
}

// For localhost we get a fake document and we only need the public key
const FakeAttestationDocumentSchema = z.object({
  public_key: z.nullable(z.instanceof(Uint8Array))
});

type FakeAttestationDocument = z.infer<typeof FakeAttestationDocumentSchema>;

async function fakeAuthenticate(
  attestationDocumentBase64: string
): Promise<FakeAttestationDocument> {
  const attestationDocumentBuffer = decode(attestationDocumentBase64);
  const cborAttestationDocument: Uint8Array[] = cbor.decode(attestationDocumentBuffer);
  const payload = cborAttestationDocument[2];
  const payloadDecoded = cbor.decode(payload);
  const zodParsed = await FakeAttestationDocumentSchema.parse(payloadDecoded);
  return zodParsed;
}

export async function verifyAttestation(nonce: string): Promise<AttestationDocument> {
  try {
    const attestationDocumentBase64 = await fetchAttestationDocument(nonce);

    // Get the API URL from the API layer where it's already set
    const apiUrl = getApiUrl();

    // With a local backend we get a fake attestation document, so we'll just pretend to authenticate it
    if (
      apiUrl &&
      (apiUrl === "http://127.0.0.1:3000" ||
        apiUrl === "http://localhost:3000" ||
        apiUrl === "http://0.0.0.0:3000")
    ) {
      console.log("DEV MODE: Using fake attestation document");
      const fakeDocument = await fakeAuthenticate(attestationDocumentBase64);
      return fakeDocument as AttestationDocument;
    }

    // The real thing!
    const verifiedDocument = await authenticate(attestationDocumentBase64, awsRootCertDer, nonce);
    return verifiedDocument;
  } catch (error) {
    if (error instanceof Error) {
      console.error("Error verifying attestation document:", error);
      throw new Error(`Couldn't process attestation document: ${error.message}`);
    } else {
      console.error("Error verifying attestation document:", error);
      throw new Error("Couldn't process attestation document.");
    }
  }
}

File: rust/src/attestation.rs
use crate::error::{Error, Result};
use base64::{engine::general_purpose::STANDARD as BASE64, Engine};
use ring::signature;
use serde::{Deserialize, Serialize};
use serde_cbor::Value as CborValue;
use x509_parser::prelude::*;

// AWS Nitro Root Certificate (production)
const AWS_NITRO_ROOT_CERT: &[u8] = include_bytes!("../assets/aws_nitro_root.der");

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AttestationDocument {
    pub module_id: String,
    pub timestamp: u64,
    pub digest: String,
    pub pcrs: std::collections::HashMap<usize, Vec<u8>>,
    pub certificate: Vec<u8>,
    pub cabundle: Vec<Vec<u8>>,
    pub public_key: Option<Vec<u8>>,
    pub user_data: Option<Vec<u8>>,
    pub nonce: Option<Vec<u8>>,
}

pub struct AttestationVerifier {
    expected_pcrs: Option<std::collections::HashMap<usize, Vec<u8>>>,
    allow_debug: bool,
}

#[allow(clippy::derivable_impls)]
impl Default for AttestationVerifier {
    fn default() -> Self {
        Self {
            expected_pcrs: None,
            allow_debug: cfg!(feature = "mock-attestation"),
        }
    }
}

impl AttestationVerifier {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_expected_pcrs(mut self, pcrs: std::collections::HashMap<usize, Vec<u8>>) -> Self {
        self.expected_pcrs = Some(pcrs);
        self
    }

    pub fn verify_attestation_document(
        &self,
        document_b64: &str,
        expected_nonce: &str,
    ) -> Result<AttestationDocument> {
        let document_bytes = BASE64.decode(document_b64)?;

        // Parse COSE_Sign1 structure
        let cbor_value: CborValue = serde_cbor::from_slice(&document_bytes).map_err(Error::Cbor)?;

        let cose_sign1 = match &cbor_value {
            CborValue::Array(arr) => arr,
            _ => {
                return Err(Error::AttestationVerificationFailed(
                    "Invalid COSE_Sign1 structure".to_string(),
                ))
            }
        };

        if cose_sign1.len() != 4 {
            return Err(Error::AttestationVerificationFailed(
                "COSE_Sign1 must have 4 elements".to_string(),
            ));
        }

        // Extract components
        let protected = match &cose_sign1[0] {
            CborValue::Bytes(b) => b,
            _ => {
                return Err(Error::AttestationVerificationFailed(
                    "Invalid protected header".to_string(),
                ))
            }
        };

        let payload = match &cose_sign1[2] {
            CborValue::Bytes(b) => b,
            _ => {
                return Err(Error::AttestationVerificationFailed(
                    "Invalid payload".to_string(),
                ))
            }
        };

        let signature = match &cose_sign1[3] {
            CborValue::Bytes(b) => b,
            _ => {
                return Err(Error::AttestationVerificationFailed(
                    "Invalid signature".to_string(),
                ))
            }
        };

        // Parse attestation document from payload
        let doc_cbor: CborValue = serde_cbor::from_slice(payload).map_err(Error::Cbor)?;

        let doc = self.parse_attestation_document(&doc_cbor)?;

        // Verify nonce
        if let Some(nonce_bytes) = &doc.nonce {
            let nonce_str = String::from_utf8(nonce_bytes.to_vec()).map_err(|e| {
                Error::AttestationVerificationFailed(format!("Invalid nonce encoding: {}", e))
            })?;

            if nonce_str != expected_nonce {
                return Err(Error::AttestationVerificationFailed(
                    "Nonce mismatch".to_string(),
                ));
            }
        } else {
            return Err(Error::AttestationVerificationFailed(
                "Missing nonce in attestation document".to_string(),
            ));
        }

        // Verify certificate chain
        self.verify_certificate_chain(&doc)?;

        // Verify signature
        self.verify_signature(protected, payload, signature, &doc)?;

        // Verify PCRs if expected
        if let Some(expected_pcrs) = &self.expected_pcrs {
            self.verify_pcrs(&doc, expected_pcrs)?;
        }

        Ok(doc)
    }

    fn parse_attestation_document(&self, cbor: &CborValue) -> Result<AttestationDocument> {
        let map = match cbor {
            CborValue::Map(m) => m,
            _ => {
                return Err(Error::AttestationVerificationFailed(
                    "Invalid attestation document format".to_string(),
                ))
            }
        };

        let mut doc = AttestationDocument {
            module_id: String::new(),
            timestamp: 0,
            digest: String::new(),
            pcrs: std::collections::HashMap::new(),
            certificate: Vec::new(),
            cabundle: Vec::new(),
            public_key: None,
            user_data: None,
            nonce: None,
        };

        for (key, value) in map {
            let key_str = match key {
                CborValue::Text(s) => s.as_str(),
                _ => {
                    return Err(Error::AttestationVerificationFailed(
                        "Invalid key in attestation document".to_string(),
                    ))
                }
            };

            match key_str {
                "module_id" => {
                    doc.module_id = match value {
                        CborValue::Text(s) => s.clone(),
                        _ => {
                            return Err(Error::AttestationVerificationFailed(
                                "Invalid module_id".to_string(),
                            ))
                        }
                    };
                }
                "timestamp" => {
                    doc.timestamp = match value {
                        CborValue::Integer(i) if *i >= 0 => *i as u64,
                        CborValue::Integer(_) => {
                            return Err(Error::AttestationVerificationFailed(
                                "Invalid timestamp: negative value".to_string(),
                            ))
                        }
                        _ => {
                            return Err(Error::AttestationVerificationFailed(
                                "Invalid timestamp: not an integer".to_string(),
                            ))
                        }
                    };
                }
                "digest" => {
                    doc.digest = match value {
                        CborValue::Text(s) => s.clone(),
                        _ => {
                            return Err(Error::AttestationVerificationFailed(
                                "Invalid digest".to_string(),
                            ))
                        }
                    };
                }
                "pcrs" => {
                    let pcrs_map = match value {
                        CborValue::Map(m) => m,
                        _ => {
                            return Err(Error::AttestationVerificationFailed(
                                "Invalid PCRs format".to_string(),
                            ))
                        }
                    };

                    for (pcr_key, pcr_value) in pcrs_map {
                        let index = match pcr_key {
                            CborValue::Integer(i) if *i >= 0 => *i as usize,
                            CborValue::Integer(_) => {
                                return Err(Error::AttestationVerificationFailed(
                                    "Invalid PCR index: negative value".to_string(),
                                ))
                            }
                            _ => {
                                return Err(Error::AttestationVerificationFailed(
                                    "Invalid PCR index: not an integer".to_string(),
                                ))
                            }
                        };

                        let pcr_bytes = match pcr_value {
                            CborValue::Bytes(b) => b.clone(),
                            _ => {
                                return Err(Error::AttestationVerificationFailed(
                                    "Invalid PCR value".to_string(),
                                ))
                            }
                        };

                        doc.pcrs.insert(index, pcr_bytes);
                    }
                }
                "certificate" => {
                    doc.certificate = match value {
                        CborValue::Bytes(b) => b.clone(),
                        _ => {
                            return Err(Error::AttestationVerificationFailed(
                                "Invalid certificate".to_string(),
                            ))
                        }
                    };
                }
                "cabundle" => {
                    let bundle = match value {
                        CborValue::Array(a) => a,
                        _ => {
                            return Err(Error::AttestationVerificationFailed(
                                "Invalid cabundle".to_string(),
                            ))
                        }
                    };

                    for cert in bundle {
                        let cert_bytes = match cert {
                            CborValue::Bytes(b) => b.clone(),
                            _ => {
                                return Err(Error::AttestationVerificationFailed(
                                    "Invalid certificate in bundle".to_string(),
                                ))
                            }
                        };
                        doc.cabundle.push(cert_bytes);
                    }
                }
                "public_key" => {
                    doc.public_key = match value {
                        CborValue::Bytes(b) => Some(b.clone()),
                        _ => None,
                    };
                }
                "user_data" => {
                    doc.user_data = match value {
                        CborValue::Bytes(b) => Some(b.clone()),
                        _ => None,
                    };
                }
                "nonce" => {
                    doc.nonce = match value {
                        CborValue::Bytes(b) => Some(b.clone()),
                        _ => None,
                    };
                }
                _ => {} // Ignore unknown fields
            }
        }

        Ok(doc)
    }

    fn verify_certificate_chain(&self, doc: &AttestationDocument) -> Result<()> {
        // In mock mode, skip certificate verification
        if self.allow_debug && doc.module_id.starts_with("mock-") {
            return Ok(());
        }

        // Step 1: Verify the first cert in cabundle matches AWS Nitro root
        if doc.cabundle.is_empty() {
            return Err(Error::AttestationVerificationFailed(
                "Certificate bundle is empty".to_string(),
            ));
        }

        if doc.cabundle[0] != AWS_NITRO_ROOT_CERT {
            return Err(Error::AttestationVerificationFailed(
                "First certificate does not match AWS Nitro root certificate".to_string(),
            ));
        }

        // Step 2: Parse all certificates and check validity
        let mut certs = Vec::new();
        for (i, cert_der) in doc.cabundle.iter().enumerate() {
            let (_, cert) = X509Certificate::from_der(cert_der).map_err(|e| {
                Error::AttestationVerificationFailed(format!(
                    "Failed to parse certificate {}: {:?}",
                    i, e
                ))
            })?;

            // Check certificate validity
            if !cert.validity().is_valid() {
                return Err(Error::AttestationVerificationFailed(format!(
                    "Certificate {} is expired or not yet valid",
                    i
                )));
            }

            certs.push(cert);
        }

        // Parse the leaf certificate
        let (_, leaf_cert) = X509Certificate::from_der(&doc.certificate).map_err(|e| {
            Error::AttestationVerificationFailed(format!(
                "Failed to parse leaf certificate: {:?}",
                e
            ))
        })?;

        if !leaf_cert.validity().is_valid() {
            return Err(Error::AttestationVerificationFailed(
                "Leaf certificate is expired or not yet valid".to_string(),
            ));
        }

        // Step 3: Verify the certificate chain signatures
        // AWS Nitro chain: root -> regional -> zonal -> instance -> leaf
        // Each cert must be signed by the PREVIOUS cert in the hierarchical chain

        // Verify each cert (except root) is signed by the previous cert
        for i in 1..certs.len() {
            let cert = &certs[i];
            let cert_der = &doc.cabundle[i];
            let issuer = &certs[i - 1]; // The issuer should be the previous cert in the chain

            // Verify the issuer/subject relationship
            if cert.issuer() != issuer.subject() {
                return Err(Error::AttestationVerificationFailed(format!(
                    "Certificate {} issuer doesn't match certificate {} subject - chain is broken",
                    i,
                    i - 1
                )));
            }

            // Verify the signature
            if !self.verify_cert_signature(cert_der, issuer)? {
                return Err(Error::AttestationVerificationFailed(format!(
                    "Certificate {} signature verification failed (not signed by certificate {})",
                    i,
                    i - 1
                )));
            }
        }

        // Verify the leaf certificate is signed by the last cert in the chain
        if !certs.is_empty() {
            let last_cert = &certs[certs.len() - 1];

            // Verify the issuer/subject relationship
            if leaf_cert.issuer() != last_cert.subject() {
                return Err(Error::AttestationVerificationFailed(
                    "Leaf certificate issuer doesn't match last certificate in chain".to_string(),
                ));
            }

            // Verify the signature
            if !self.verify_cert_signature(&doc.certificate, last_cert)? {
                return Err(Error::AttestationVerificationFailed(
                    "Leaf certificate signature verification failed".to_string(),
                ));
            }
        }

        Ok(())
    }

    fn verify_cert_signature(&self, cert_der: &[u8], issuer: &X509Certificate) -> Result<bool> {
        // Parse the certificate to get its TBS (to-be-signed) portion and signature
        let (_, cert) = X509Certificate::from_der(cert_der).map_err(|e| {
            Error::AttestationVerificationFailed(format!(
                "Failed to parse certificate for signature verification: {:?}",
                e
            ))
        })?;

        // The signature algorithm should match what AWS Nitro uses
        let sig_algo = &cert.signature_algorithm;
        let sig_oid = sig_algo.algorithm.to_id_string();

        // AWS Nitro uses ECDSA with P-384 and SHA-384 (OID: 1.2.840.10045.4.3.3)
        if sig_oid != "1.2.840.10045.4.3.3" {
            // Also support P-256 with SHA-256 (OID: 1.2.840.10045.4.3.2) for compatibility
            if sig_oid != "1.2.840.10045.4.3.2" {
                return Ok(false); // Unsupported algorithm
            }
        }

        // Extract the issuer's public key
        let issuer_pubkey = issuer.public_key();

        // The public key is in SubjectPublicKeyInfo format
        // For EC keys, we need to extract the actual EC point
        let pubkey_bytes = issuer_pubkey.raw;

        // Find the EC point in the public key data
        // EC points start with 0x04 (uncompressed) and are 97 bytes for P-384, 65 for P-256
        let ec_point = if sig_oid == "1.2.840.10045.4.3.3" {
            // P-384: 97 bytes (0x04 + 48 bytes X + 48 bytes Y)
            extract_ec_point(pubkey_bytes, 97)
        } else {
            // P-256: 65 bytes (0x04 + 32 bytes X + 32 bytes Y)
            extract_ec_point(pubkey_bytes, 65)
        }?;

        // Get the TBS certificate data and signature
        let tbs_cert = cert.tbs_certificate.as_ref();
        let signature = cert.signature_value.as_ref();

        // Verify the signature using ring
        let verification_alg = if sig_oid == "1.2.840.10045.4.3.3" {
            &signature::ECDSA_P384_SHA384_ASN1
        } else {
            &signature::ECDSA_P256_SHA256_ASN1
        };

        let public_key = signature::UnparsedPublicKey::new(verification_alg, ec_point);

        public_key
            .verify(tbs_cert, signature)
            .map(|_| true)
            .map_err(|_| {
                Error::AttestationVerificationFailed(
                    "Certificate signature verification failed".to_string(),
                )
            })
    }

    fn verify_signature(
        &self,
        protected: &[u8],
        payload: &[u8],
        signature_bytes: &[u8],
        doc: &AttestationDocument,
    ) -> Result<()> {
        // In mock mode, skip signature verification
        if self.allow_debug && doc.module_id.starts_with("mock-") {
            return Ok(());
        }

        // Parse the leaf certificate
        let (_, cert) = X509Certificate::from_der(&doc.certificate).map_err(|e| {
            Error::AttestationVerificationFailed(format!(
                "Failed to parse leaf certificate: {:?}",
                e
            ))
        })?;

        // Extract the public key bytes from the certificate
        let public_key_info = cert.public_key();

        // AWS Nitro uses P-384, extract the EC point properly
        // P-384: 97 bytes (0x04 + 48 bytes X + 48 bytes Y)
        let public_key_bytes = extract_ec_point(public_key_info.raw, 97)?;

        // Create the COSE_Sign1 signature structure
        // This follows the COSE specification for the data to be signed
        let sig_structure = create_sig_structure(protected, payload)?;

        // For ECDSA P-384 with SHA-384 (which is what AWS Nitro uses)
        // AWS Nitro uses raw signatures (r||s), not ASN.1 encoded
        let public_key = signature::UnparsedPublicKey::new(
            &signature::ECDSA_P384_SHA384_FIXED,
            public_key_bytes,
        );

        // Verify the signature
        public_key
            .verify(&sig_structure, signature_bytes)
            .map_err(|_| {
                Error::AttestationVerificationFailed("Signature verification failed".to_string())
            })?;

        Ok(())
    }

    fn verify_pcrs(
        &self,
        doc: &AttestationDocument,
        expected: &std::collections::HashMap<usize, Vec<u8>>,
    ) -> Result<()> {
        for (index, expected_value) in expected {
            match doc.pcrs.get(index) {
                Some(actual_value) => {
                    if actual_value != expected_value {
                        return Err(Error::AttestationVerificationFailed(format!(
                            "PCR{} mismatch",
                            index
                        )));
                    }
                }
                None => {
                    return Err(Error::AttestationVerificationFailed(format!(
                        "PCR{} missing",
                        index
                    )));
                }
            }
        }
        Ok(())
    }
}

fn extract_ec_point(pubkey_bytes: &[u8], expected_size: usize) -> Result<&[u8]> {
    // The public key is in SubjectPublicKeyInfo format (ASN.1 DER encoded)
    // We need to extract the actual EC point from the BIT STRING

    // The structure is:
    // SEQUENCE {
    //   algorithm AlgorithmIdentifier,
    //   subjectPublicKey BIT STRING
    // }

    // For EC keys, x509-parser gives us the raw bytes which includes the full
    // SubjectPublicKeyInfo structure. The EC point is at the end after the
    // algorithm identifier and is preceded by a BIT STRING tag.

    // Look for BIT STRING tag (0x03) followed by length and unused bits (0x00)
    // The EC point follows immediately after
    for i in 0..pubkey_bytes.len() {
        if pubkey_bytes[i] == 0x03 {
            // BIT STRING tag
            if i + 2 < pubkey_bytes.len() {
                // Next byte is length (for EC keys, usually 0x42 for P-256 or 0x62 for P-384)
                // Then 0x00 for no unused bits
                // Then 0x04 for uncompressed point
                if i + 3 < pubkey_bytes.len()
                    && pubkey_bytes[i + 2] == 0x00
                    && pubkey_bytes[i + 3] == 0x04
                {
                    let ec_point_start = i + 3;
                    let remaining = &pubkey_bytes[ec_point_start..];
                    if remaining.len() == expected_size {
                        return Ok(remaining);
                    }
                }
            }
        }
    }

    // Fallback: The x509-parser might have already extracted just the key material
    // In this case, look for the uncompressed point marker (0x04) at the expected position
    if pubkey_bytes.len() >= expected_size {
        // Try from the end (most common case with x509-parser)
        let from_end = &pubkey_bytes[pubkey_bytes.len() - expected_size..];
        if from_end[0] == 0x04 {
            return Ok(from_end);
        }

        // Try from a typical offset (after algorithm OID and parameters)
        // For EC keys, this is often around offset 23-27
        for offset in [23, 24, 25, 26, 27].iter() {
            if *offset + expected_size <= pubkey_bytes.len() {
                let candidate = &pubkey_bytes[*offset..*offset + expected_size];
                if candidate[0] == 0x04 {
                    return Ok(candidate);
                }
            }
        }
    }

    Err(Error::AttestationVerificationFailed(format!(
        "Failed to extract EC public key point (expected {} bytes, pubkey is {} bytes)",
        expected_size,
        pubkey_bytes.len()
    )))
}

fn create_sig_structure(protected: &[u8], payload: &[u8]) -> Result<Vec<u8>> {
    // Create the COSE_Sign1 signature structure as a CBOR array
    // ["Signature1", protected, external_aad, payload]
    let sig_structure = CborValue::Array(vec![
        CborValue::Text("Signature1".to_string()),
        CborValue::Bytes(protected.to_vec()),
        CborValue::Bytes(vec![]), // empty external AAD
        CborValue::Bytes(payload.to_vec()),
    ]);

    // Encode to CBOR bytes
    serde_cbor::to_vec(&sig_structure).map_err(Error::Cbor)
}

#[cfg(feature = "mock-attestation")]
pub fn create_mock_attestation_document(nonce: &str) -> Result<String> {
    use serde_cbor::to_vec;
    use std::collections::HashMap;

    let mut pcrs = HashMap::new();
    pcrs.insert(0, vec![0u8; 48]); // Mock PCR0

    let doc = AttestationDocument {
        module_id: "mock-module".to_string(),
        timestamp: chrono::Utc::now().timestamp() as u64,
        digest: "SHA384".to_string(),
        pcrs,
        certificate: vec![0u8; 256],    // Mock certificate
        cabundle: vec![vec![0u8; 256]], // Mock CA bundle
        public_key: None,
        user_data: None,
        nonce: Some(nonce.as_bytes().to_vec()),
    };

    // Create a mock COSE_Sign1 structure
    let payload = to_vec(&doc).map_err(Error::Cbor)?;
    let protected = vec![0u8; 32]; // Mock protected header
    let signature = vec![0u8; 64]; // Mock signature

    let cose_sign1 = vec![
        CborValue::Bytes(protected),
        CborValue::Map(std::collections::BTreeMap::new()), // Empty unprotected headers
        CborValue::Bytes(payload),
        CborValue::Bytes(signature),
    ];

    let cose_bytes = to_vec(&CborValue::Array(cose_sign1)).map_err(Error::Cbor)?;
    Ok(BASE64.encode(cose_bytes))
}
